packages/langium/src/grammar/type-system/ast-collector.ts

```diff
@@ -20,21 +20,18 @@ export function collectAst(grammars: Grammar | Grammar[], documents?: LangiumDoc
     const { inferred, declared } = collectTypeResources(grammars, documents);
 
     const astTypes = {
-        interfaces: mergeAndRemoveDuplicates<InterfaceType>(inferred.interfaces,declared.interfaces),
+        interfaces: sortInterfacesTopologically(mergeAndRemoveDuplicates<InterfaceType>(inferred.interfaces, declared.interfaces)),
         unions: mergeAndRemoveDuplicates<UnionType>(inferred.unions, declared.unions),
     };
 
-    sortInterfacesTopologically(astTypes.interfaces);
-    astTypes.unions.sort((a, b) => a.name.localeCompare(b.name));
-
     specifyAstNodeProperties(astTypes);
     return astTypes;
 }
 
 function mergeAndRemoveDuplicates<T extends { name: string }>(inferred: T[], declared: T[]): T[] {
     return Array.from(inferred.concat(declared)
         .reduce((acc, type) => { acc.set(type.name, type); return acc; }, new Map<string, T>())
-        .values());
+        .values()).sort((a, b) => a.name.localeCompare(b.name));
```

Comment:
Interfaces do not need this sorting -- it's only for unions. I'd suggest applying it directly to unions like `sortInterfacesTopologically`.

Reply:
Interfaces also benefit from this, since it keeps interfaces stable in case the grammar changes. The topological sorting function is stable and so we get less changes in our output files.